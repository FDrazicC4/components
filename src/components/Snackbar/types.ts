import React from 'react';
import { Colors } from '@components/types';


export interface Snackbar {
    position: 'top-right' | 'top-center' | 'top-left' | 'bottom-right' | 'bottom-center' | 'bottom-left',
    message: React.ReactNode,
    key?: React.Key,
    duration?: number,
    color: Colors,
    open: boolean,
    otherProps: any,
}


export interface PositionsObject {
    'top-right'?: Snackbar[],
    'top-left'?: Snackbar[],
    'top-center'?: Snackbar[],
    'bottom-right'?: Snackbar[],
    'bottom-center'?: Snackbar[],
    'bottom-left'?: Snackbar[],
}
