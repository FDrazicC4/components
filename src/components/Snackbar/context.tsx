import React from 'react';


interface ContextType {
    enqueue: (enqueue: any) => void,
    remove: (key: React.Key) => void,
}


const SnackbarContext = React.createContext<Partial<ContextType>>({});


export default SnackbarContext;
