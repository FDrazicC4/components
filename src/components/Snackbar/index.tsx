import React from 'react';
import clsx from 'clsx';
import { Snackbar as MuiSnackbar, SnackbarContent as MuiSnackbarContent } from '@material-ui/core';


const Snackbar = React.forwardRef(({
    action,
    className,
    children,
    color,
    duration = 0,
    message,
    onClose,
    open,
    position = 'top-right',
    ...otherProps
}: any, ref: React.Ref<any>): React.ReactElement => {


    const getAnchor = (): any => {

        const anchorArray = position.split('-');

        const [vertical, horizontal] = anchorArray;

        return {
            vertical,
            horizontal,
        };
    };


    const snackbarClasses = clsx(
        'brew-Snackbar',
        `brew-Snackbar--${color}`,
        className,
    );


    const contentClasses = {
        message: 'brew-Snackbar__message',
        action: 'brew-Snackbar__action',
    };


    return (
        <MuiSnackbar
            action={action}
            anchorOrigin={getAnchor()}
            autoHideDuration={duration * 1000 || null}
            className={children ? snackbarClasses : ''}
            message={message}
            onClose={onClose}
            open={open}
            ref={ref}
            {...otherProps}
        >
            {children || (
                <MuiSnackbarContent
                    action={action}
                    classes={contentClasses}
                    className={snackbarClasses}
                    message={message}
                />
            )}
        </MuiSnackbar>
    );


});


Snackbar.displayName = 'Snackbar';


export default Snackbar;
