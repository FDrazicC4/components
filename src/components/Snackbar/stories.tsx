import React from 'react';
import { button, number, select, text } from '@storybook/addon-knobs';
import Snackbar from './index';


export default {
    component: Snackbar,
    title: 'Development|Snackbar',
};


export const Sandbox = (): React.ReactElement => {


    const [open, setOpen] = React.useState(true);


    const position = select(
        'position',
        ['top-right', 'top-center', 'top-left', 'bottom-right', 'bottom-center', 'bottom-left'],
        'top-right',
    );


    const onClose = (event: React.SyntheticEvent, reason: string): void => {
        if (event) event.stopPropagation();
        if (reason === 'clickaway') return;

        setOpen(false);
    };


    const dismissSnackbar = (): void => setOpen(false);


    const colors = select(
        'color',
        ['primary', 'secondary', 'tertiary', 'success', 'error', 'warning', 'info', 'light', 'dark'],
        'primary',
    );


    const message = text('message', 'Message Default');
    const duration = number('duration', 0);
    const iconName = text('iconName', '');


    button('toggle', () => setOpen((prev: any) => !prev));


    return (
        <Snackbar
            action={<div onClick={dismissSnackbar}>Close</div>}
            color={colors}
            duration={duration}
            iconName={iconName}
            isOpen={open}
            message={message}
            onClose={onClose}
            position={position}
        />
    );


};


export const Uncontrolled = (): React.ReactElement => (
    <Snackbar
        action="close"
        duration={5000}
        message="Foo Bar"
        slideDirection="up"
    />
);
