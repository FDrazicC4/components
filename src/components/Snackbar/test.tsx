import React from 'react';
import renderer from 'react-test-renderer';
import Snackbar from './index';


it('renders correctly', () => {


    const tree = renderer
        .create(
            <Snackbar
                action={<div onClick={() => console.log('bar')}>Close</div>}
                color="primary"
                isOpen
                message="foo"
                onClose={() => console.log('foo')}
                position="top-right"
            />,
            {
                createNodeMock: node => document.createElement(node.type as any),
            }).toJSON();


    expect(tree).toMatchSnapshot();


});
