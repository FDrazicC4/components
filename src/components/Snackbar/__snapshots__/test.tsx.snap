// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="MuiSnackbar-root brew-Snackbar brew-Snackbar--primary brew-Snackbar--top-right MuiSnackbar-anchorOriginTopRight"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="brew-Snackbar__content"
    style={
      Object {
        "visibility": undefined,
      }
    }
  >
    <div
      className="brew-Snackbar__message"
    >
      <span
        className="MuiTypography-root brew-Typography brew-Typography--display-initial brew-Typography--variant-caption brew-Typography--align-inherit brew-Typography--color-inherit MuiTypography-caption"
      >
        foo
      </span>
    </div>
    <span
      className="brew-Snackbar__action"
      onClick={[Function]}
    >
      <span
        className="MuiTypography-root brew-Typography brew-Typography--display-initial brew-Typography--variant-caption brew-Typography--align-inherit brew-Typography--color-inherit MuiTypography-caption"
      >
        <div
          onClick={[Function]}
        >
          Close
        </div>
      </span>
    </span>
  </div>
</div>
`;
